transmission.go

```diff
@@ -145,68 +148,84 @@ func respondErr(e *encoder, handle uint64, err error) {
 	rep.encode(e)
 }
 
-// ctxRW wraps a net.Conn to provide cancellation. It does so by setting a
-// low-ish timeout on each read/write call. If the call times out, it checks if
-// its context is done and if not, retries it.
+// rwRequest is a request to execute a Read or Write call with a given buffer.
+type rwRequest struct {
+	call func(p []byte) (n int, err error)
+	p    []byte
+}
+
+// rwResponse contains the return values from the readOrWrite function in the
+// corresponding rwRequest, or the context error if the context was cancelled.
+type rwResponse struct {
+	n   int
+	err error
+}
+
+// ctxRW wraps a net.Conn to respect context cancellation. It does so by
+// fulfilling reads and writes in a background goroutine, and returning when
+// either the goroutine has completed the request or the context is cancelled.
 type ctxRW struct {
-	ctx   context.Context
-	c     net.Conn
-	hasDL bool
-	dl    time.Time
+	ctx          context.Context
+	c            net.Conn
+	requestChan  chan rwRequest
+	responseChan chan rwResponse
+	closeChan    chan struct{}
 }
 
 // wrapConn wraps a connection in a ctxRW.
-func wrapConn(ctx context.Context, c net.Conn) io.ReadWriter {
+func wrapConn(ctx context.Context, c net.Conn) io.ReadWriteCloser {
 	dl, ok := ctx.Deadline()
-	return &ctxRW{ctx, c, ok, dl}
+	requestChan := make(chan rwRequest, 1)
+	resultChan := make(chan rwResponse, 1)
+	closeChan := make(chan struct{})
+	// Start read/write worker.
+	go func() {
+		for {
+			select {
+			case req := <-requestChan:
+				n, err := req.call(req.p)
+				resultChan <- rwResponse{n, err}
+			case <-ctx.Done():
+				return
+			case <-closeChan:
+				return
+			}
+		}
+	}()
+	return &ctxRW{ctx, c, ok, dl, requestChan, resultChan, closeChan}
 }
 
-// maybeIgnore checks whether err is an error we want to ignore (i.e. a timeout
-// without ctx being cancelled) and returns nil, if so.
-func (rw *ctxRW) maybeIgnore(err error) error {
-	if e := rw.ctx.Err(); e != nil {
-		return e
-	}
-	if to, ok := err.(interface{ Timeout() bool }); ok && to.Timeout() {
-		return nil
-	}
-	return err
-}
+// do sends a Read or Write request to the background worker, returning when the
+// operation is completed or the context is cancelled.
+func (rw *ctxRW) do(op rwRequest) (n int, err error) {
+	// Note: if the ctxRW is already closed here, the op will just be buffered.
+	rw.requestChan <- op
 
-// setDeadline sets the deadline for the next read/write.
-func (rw *ctxRW) setDeadline() {
-	dl := time.Now().Add(100 * time.Millisecond)
```

Comment:
I wasn't sure whether this 100ms polling was serving another purpose beyond just allowing a ctx cancellation check every so often (e.g. not sure if this doubles as some sort of keepalive ping?) but if this was purely for cancellation I think the new approach works (albeit at the cost of having to call `Close()`, which seems worth it to me?)

Separately, another approach I was considering was just making this change in [Loopback](https://github.com/Merovius/nbd/blob/7ffd8729581328cb2ff32d8dc52d684f272f31b9/netlink.go#L81-L101):

```
eg.Go(func() {
  <-ctx.Done()
  serverc.Close() // moved up from the other eg.Go()
})
```

But I figured that the other users of `wrapConn` may run into the same issue, so it seemed to make sense to fix it directly in `ctxRW`.

Also, thank you for all the help so far looking at issues / PRs - do let me know if this is taking up too much of your time, I am happy to make this change as a patch or fork instead.

Reply:
Just as a very first thought: The "best" way to abort a Read/Write on a `net.Conn` is to call [SetDeadline](https://pkg.go.dev/net#Conn.SetDeadline) with a time in the past. I think a sequence for making a read respect cancellation [looks something like this](https://go.dev/play/p/JYiJGsb48aH) (untested, just off the top of my head). This makes the read have a bunch of overhead, so depending on the context (no pun intended) it might make sense to do this at a higher level (e.g. if a single, cancelable operation needs a bunch of reads, it would make sense to do this wrapping at that operation level).

I didn't know about this possibility (of using `Set*Deadline`) when I wrote this package, so I didn't use it. And perhaps the abstractions already in the code aren't conducive to it. But ultimately, it would probably be best if this was the mechanism used to implement cancellation.

I will still have a look at this PR if you want (later today or over the weekend), but perhaps you'd like to incorporate this idea yourself? FWIW I do genuinely enjoy being prompted to work on this, so keep it coming.

Reply:
Looking at the code, ISTM that `wrapConn` would be the place to do the context-wrapping and spawning of the goroutine for cancellation. I think it would allow to get rid of most of the channel-logic (in particular, `Read`/`Write` could just directly delegate to the `net.Conn`), making it *significantly* easier to reason about.

Reply:
Wow, that's a nice trick! The code is much simpler now. It passes our test suite so I do believe it's working. As a sanity check, I tried removing the `SetDeadline` call, and the test hangs while trying to disconnect the NBD, as expected.

BTW, I force-pushed the changes so that this PR could be kept as a single commit, but I know some people prefer making changes as commits then squashing when merging into main. Let me know if you have any preferences.

Reply:
No, I strongly prefer 1. every PR to be a single commit and 2. to have the final version history be a clean, individually buildable and understandable change. So, all perfect.

I'll have a look at the change.